### DISCUSSION TITLE
Discuss the new generation of programming languages, the basic features of V, such as syntax simplification, code input convenience,V、Rust、Go etc.

### DISCUSSION THREAD

**Discussion by X1aSheng** [vlang/v#14234] [General]
讨论新一代编程语言Vlang的相关基本特性，如语法简化、代码输入便捷化，V、Rust、Go、Koltlin对比。
Discuss the new generation of programming languages, the basic features of V, such as syntax simplification, code input convenience, etc.

V:
```
fn main() {
    areas := ['game', 'web', 'tools', 'science', 'systems',
              'embedded', 'drivers', 'GUI', 'mobile']
    for area in areas {
        println('Hello, $area developers!')
    }
}
```
Rust:
```
fn main() {
let a = [10, 20, 30, 40, 50];
    for i in 0..5 {
        println!("a[{}] = {}", i, a[i]);
    }
}
```
Go:
```
package main

import "fmt"

func main() {
        sum := 0
        for i := 0; i <= 10; i++ {
                sum += i
        }
        fmt.Println(sum)
}
```

**Comment by fxyc87** [comment]
V语言确实不错，加油。

**Comment by Wajinn** [comment]
Seems a bit too general.  In English, it appears you are telling or pushing others to discuss a topic.  It might be better for the OP to start the discussion, and to be more specific about the differences in arrays or the purpose of the comparison.

Also, V contains each type of for-loop shown.  So the for-loops shown for Go and Rust, both can be done in V.

**Comment by JalonSolov** [comment]
https://github.com/vlang/v/blob/master/doc/docs.md#for-loop

**Comment by IngwiePhoenix** [comment]
I love the majority of V's syntax, but there are missed opportunities as well as some parts that I would actually describe as "cringe" - or, more formally, "unnessessary".

A friend of mine wanted to see some V code so I just threw something together:

```v
[baseUrl: "/_matrix"]
struct MatrixController {}

["get"]
[header: "Cache-control: none"]
fn(m MatrixController) sync() {}

fn main() {
    mut url := ""
    mut mt_attrs := map[string][]string{}
    $for a in MatrixController.attributes {
        if a.has_arg && a.name == "baseUrl" {
            url = a.arg
        }
    }
    $for m in MatrixController.methods {
        mt_attrs[m.name] = m.attrs
    }
    //println(url) // /_matrix
    //println(mt_attrs) // {'sync': ['get']}

    // make some silly URLs
    for m_name, _ in mt_attrs {
        println("URL: " + [url, m_name].join("/"))
    }
}
```

Not `v fmt`'d or anything, just some raw "write it down to share".

- Attributes for structs and methods are implemented differently.
  * Structs have their own `StructAttribute` in `builtin`
  * Methods are only an array of strings
- Callbacks... possibly my most favorite rant topic in regards to V. Sorry in advance. ^^;

Take this for instance:
```v
type doSomething_cb = fn(int,int)int
fn doSomething(cb doSomething_cb) {
  println( cb(2, 2) )
}

doSomething(fn(a int, b int) int {
  return a + b
})
```

Bit of an odd example, yes. But as you can see, you have to write the entire signature for a callback, although it is already known way in advance.

The third thing that I am still not too happy with for V is it's syntax magic. Some methods and features rely heavily on compiler magic implemented in the compiler that, unless you lock yourself into only using the V standard library for everything, you can not implement for something else or your own purpose.

Take the above example. When sorting an array or mapping through, you don't even have to specify the whole callback, but can just write the predicate as-is like `arr.sort(x > 10)`. Neat, but only works on built-in types. I can't write my example as `doSomething(a + b)`, not even `doSomething(fn(a, b) [ return a + b})`.

That said, I love V. Maybe eventually those quirks get worked on and go away =)

If you like V, you might actually want to look into D and Swift, its worth a peek!

**Reply by medvednikov** [reply]
There's no short syntax for callbacks to follow the one way rule. 

But I agree it'd be nice, especially since V already figures out the type in stuff like `mut x := ...`.

Perhaps something can be arranged.

**Reply by spytheman** [reply]
ok, lets take your example and actually make it compile in v:
```v
type CB = fn (int, int) int

fn do_something(cb CB) {
        println(cb(2, 2))
}

do_something(fn (a int, b int) int {
        return a + b
})
```

Now, lets simplify it, while still keeping it compilable:
```v
fn do_something(cb fn (int, int) int) {
	println(cb(2, 2))
}

do_something(fn (a int, b int) int {
	return a + b
})
```

I really would like for people to actually keep up with V, and to have educated opinions on it, instead of judging it by their faint memories, or worse, by *other people*'s faint memories.

**Reply by spytheman** [reply]
> When sorting an array or mapping through, you don't even have to specify the whole callback, but can just write the predicate as-is like arr.sort(x > 10). Neat, but only works on built-in types.

That is *still* true, and I do not like that this magic is not general too. I do not know how to make it general, without introducing support for macros :-| .

**Reply by Wajinn** [reply]
> I really would like for people to actually keep up with V, and to have educated opinions on it, instead of judging it by their faint memories, or worse, by _other people_'s faint memories.

I've noticed that too, where some people will base their opinion about V, based on somebody else's opinion that can even be from years ago.  Hopefully, as V gets to 0.3, there will be more people developing their expertise in it.

Personally, I like V's syntax and the structure of the language.  I don't like putting the names of other languages out there, because I'm hesitant to put more gasoline on the usual fires of language wars.  But, in terms of syntax and the structure of the language, I strongly believe that V has hit the mark and has the upper hand on a lot of rivals.  

Partly because as a newer language, V was able to see some mistakes that other languages made or trapped themselves into, but also I think this is due to the vision of the core developers and contributors.  When I compare V, it comes off as significantly easier to understand and write.  I hope it stays that way, because it always seems like there are those that wish to bring in contrived complexity and unnecessary confusion, versus greater simplicity or practicality.  I wouldn't want V to run itself in circles, in great academic debates over minor things, when it can be of greater usability instead.

**Reply by shelby3** [reply]
> > When sorting an array or mapping through, you don't even have to specify the whole callback, but can just write the predicate as-is like arr.sort(x > 10). Neat, but only works on built-in types.
> 
> That is _still_ true, and I do not like that this magic is not general too. I do not know how to make it general, without introducing support for macros :-| .

Scala allows using an underscore in general to mean the first parameter of the presumed but omitted function that contains the expression. Double underscore could be any second parameter.

So you’d write generally `arr.sort(_ > 10)` without the need for macros.

> There's no short syntax for callbacks to follow the one way rule.
>
> But I agree it'd be nice

Imho, a short syntax for callbacks wouldn’t break the one way rule in the sense that short-syntax can’t already be done generally. Imho as long as there are not multiple ways to write long and/or multiple ways to write short syntax for callbacks, thus there would be only one syntax for each.

**Comment by shelby3** [comment]
The design choices for V compared to Go which I think are quite significant:

1. Avoided the inane conflation of capitalization and exporting names. I want to use capitalization for type names not functions and methods. How else am I to efficiently differentiate type names in my program if subjected to Go’s conflation?

2. Enforced [immutability](https://github.com/vlang/v/blob/master/doc/docs.md#pure-functions-by-default).

3. Apparently V has only [compile-time](https://github.com/vlang/v/blob/master/doc/docs.md#compile-time-reflection) and not runtime reflection? Should the following be added to the comparison with Go on the website? I wrote the following on May 20, 2018:

   > GopherJS [needs more work](https://github.com/gopherjs/gopherjs/issues/136#issuecomment-390473355) on dead code elimination (DCE) and unpacking to multiple files. DCE is [hampered significantly](https://github.com/gopherjs/gopherjs/issues/186) by the presence of RTTI (aka reflection)— which is a PL feature that both @keean [and I](https://github.com/keean/zenscript/issues/39#issuecomment-366057703) think [is an anti-pattern](https://github.com/keean/zenscript/issues/39#issuecomment-388657011). Yet Go currently needs reflection in order to simulate <s>generics</s>[[polymorphism and abstraction](https://jozefg.bitbucket.io/posts/2013-08-23-leaving-go.html)] (very poorly and as an anti-pattern). Scala.js doesn’t have this reflection problem because [it’s disallowed](https://www.reddit.com/r/functionalprogramming/comments/7vveeu/purescript_vs_reasonmlbucklescript_in_2018/dtw35pb/).
   
   Even [must use runtime reflection](https://codefibershq.com/blog/golang-why-nil-is-not-always-nil) in Go to properly test `nil` with interfaces which otherwise behaves bizarrely.

4. > Go [duplicated the “billion dollar mistake”](https://news.ycombinator.com/item?id=12427069) in that every type can by default be nullable. There’s [no non-nullable types](https://www.reddit.com/r/golang/comments/6d6uo2/is_go_really_as_awful_as_they_say/).
   
 5. The [option](https://github.com/vlang/v/blob/master/doc/docs.md#memory-management) to eliminate the GC. A single GC for all M:N coroutines doesn’t scale well with a centralized GC because all threads must be impacted either by pauses and/or amortized costs. This was something I was eventually going to have to figure out how to change on Go if I decided to code with Go.
 
 6. [Sum types](https://github.com/vlang/v/blob/master/doc/docs.md#sum-types) and [unions](https://github.com/vlang/v/blob/master/doc/docs.md#unions). I read something about sum types not playing well with generics and I don’t yet fully understand the issue.
 
 7. “Zero cost C interop.”
 
 Features of Go missing from V:
 
 1. [Recover](https://go.dev/blog/defer-panic-and-recover) from panic.
 
 P.S. note also [my suggestion 7 hours earlier](https://github.com/vlang/v/discussions/14234#discussioncomment-2849612) which is hidden by default when you load this page.

**Reply by JalonSolov** [reply]
Recover from panic is planned, just not implemented yet.

**Reply by Wajinn** [reply]
> The design choices for V compared to Go which I think are quite significant:

In my opinion, V has made very significant and practical improvements to become a more comfortable language to use.  V has strongly differentiated itself on various features and possibilities that Go kind of started the direction to go down, but never quite got there or strangely didn't do what seemed obvious or practical.  

Some others choices that stand out are enums (https://github.com/vlang/v/blob/master/doc/docs.md#enums) and variable shadowing (https://github.com/vlang/v/blob/master/doc/docs.md#declaration-errors).  I'm really not quite sure why it was so difficult for them to actually introduce real enums into the language.  Somewhat like generics for those in Go circles (which was previously a strong issue), I've seen a bit of debate about this, where it causes lots of head scratching.  The issue of shadowing (https://en.wikipedia.org/wiki/Variable_shadowing) is a nest of annoying errors by itself, that it appears Go is simply stuck with.

> 6. [Sum types](https://github.com/vlang/v/blob/master/doc/docs.md#sum-types) and [unions](https://github.com/vlang/v/blob/master/doc/docs.md#unions). I read something about sum types not playing well with generics and I don’t yet fully understand the issue.

Not quite sure what the issue with sum types and generics is, nor have seen any posted code.  It could possibly be about something odd that a particular programmer was/is doing.  Have not run into any problems with mixing sum types and generics together, yet, to include mixing with variadics (maybe that was also needed for them).  Which by the way, think the V implementation of such features works smoothly together.  

That Go overlooked sum types, is another one of those things that seems to have hurt them a bit (in terms of practicality and usability), and is something quite fundamental, where it seems too late for them to try and awkwardly "shoehorn" into the language at this point.
